### PR TITLE
add(delivery service factory)

### PR DESCRIPTION
**Pull request by ChrOertlin**
### Added

-

### Changed

-

### Fixed

-

### How to prepare for test

- [ ] Ssh to relevant server (depending on type of change)
- [ ] Use stage: `us`
- [ ] Paxa the environment: `paxa`
- [ ] Install on stage (example for Hasta):
    ```shell
    bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-tool-stage.sh -e S_cg -t cg -b [THIS-BRANCH-NAME] -a
    ```

### How to test

- [ ] Do ...

### Expected test outcome

- [ ] Check that ...
- [ ] Take a screenshot and attach or copy/paste the output.

## Review

- [ ] Tests executed by
- [ ] "Merge and deploy" approved by
Thanks for filling in who performed the code review and the test!

### This [version](https://semver.org/) is a

- [ ] **MAJOR** - when you make incompatible API changes
- [ ] **MINOR** - when you add functionality in a backwards compatible manner
- [ ] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

## Implementation Plan

- [ ] Document in ...
- [ ] Deploy this branch on ...
- [ ] Inform to ...

### CODE REVIEW

File: cg/services/file_delivery/deliver_files_service/deliver_files_service_factory.py
Comment (karlnyr) [inline cg/services/file_delivery/deliver_files_service/deliver_files_service_factory.py:1]: delivery_service_builder.py?

File: cg/services/file_delivery/file_formatter_service/delivery_file_formatter.py [hunk 1]
```diff
@@ -1,7 +1,9 @@
 import os
 from pathlib import Path
 from cg.constants.delivery import INBOX_NAME
+
 from cg.services.file_delivery.fetch_file_service.models import DeliveryFiles, SampleFile, CaseFile
+
```
Comment (ChrOertlin) [inline cg/services/file_delivery/file_formatter_service/delivery_file_formatter.py:6]:
```suggestion
```

File: cg/services/file_delivery/file_formatter_service/delivery_file_formatter.py [hunk 2]
```diff
@@ -1,7 +1,9 @@
 import os
 from pathlib import Path
 from cg.constants.delivery import INBOX_NAME
+
```
Comment (ChrOertlin) [inline cg/services/file_delivery/file_formatter_service/delivery_file_formatter.py:4]:
```suggestion
```

File: cg/services/file_delivery/fetch_file_service/fetch_fastq_analysis_files_service.py [hunk 3]
```diff
@@ -17,30 +20,33 @@
...
         )
+
+    def _get_case(self, case_id: str) -> Case:
+        """Fetch the case from the database."""
+        return self.status_db.get_case_by_internal_id(internal_id=case_id)
+
+    def _fetch_files(self, service_class: type, case_id: str) -> DeliveryFiles:
+        """Fetch files using the provided service class."""
+        service = service_class(self.status_db, self.hk_api, tags_fetcher=self.tags_fetcher)
+        return service.get_files_to_deliver(case_id)
```
Comment (ChrOertlin) [inline cg/services/file_delivery/fetch_file_service/fetch_fastq_analysis_files_service.py:52]: Simplification that allows a standardised constructor of the class

File: cg/services/file_delivery/deliver_files_service/deliver_files_service_factory.py [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+        file_fetcher: FetchDeliveryFilesService = self._get_file_fetcher(delivery_type)
+        sample_file_formatter: SampleFileFormatter | SampleFileConcatenationFormatter = (
+            self._get_sample_file_formatter(workflow)
+        )
+        file_formatter: DeliveryFileFormattingService = DeliveryFileFormatter(
+            case_file_formatter=CaseFileFormatter(), sample_file_formatter=sample_file_formatter
+        )
+        return DeliverFilesService(
+            delivery_file_manager_service=file_fetcher,
+            move_file_service=MoveDeliveryFilesService(),
```
Comment (islean) [inline cg/services/file_delivery/deliver_files_service/deliver_files_service_factory.py:104]: Do we need to provide the move_file_service in the instantiation of the DeliverFilesService or will it always be a freshly created MoveDeliveryFilesService?

File: cg/services/file_delivery/fetch_file_service/fetch_fastq_analysis_files_service.py [hunk 5]
```diff
@@ -17,30 +20,33 @@
...
-            self.status_db,
-            self.hk_api,
-            tags_fetcher=FetchSampleAndCaseDeliveryFileTagsService(),
-        )
-
-        fastq_files: DeliveryFiles = fetch_fastq_service.get_files_to_deliver(case_id)
-        analysis_files: DeliveryFiles = fetch_analysis_service.get_files_to_deliver(case_id)
+        case = self._get_case(case_id)
+        fastq_files = self._fetch_files(FetchFastqDeliveryFilesService, case_id)
+        analysis_files = self._fetch_files(FetchAnalysisDeliveryFilesService, case_id)
```
[the remark this answers is not in the document]
Comment (islean) [reply cg/services/file_delivery/fetch_file_service/fetch_fastq_analysis_files_service.py:34]: And maybe type hints for convention

**Review by islean — APPROVED** [review]
Looks good to me!

File: cg/services/file_delivery/deliver_files_service/deliver_files_service_factory.py (hunk 4 above)
Comment (ChrOertlin) [reply cg/services/file_delivery/deliver_files_service/deliver_files_service_factory.py:104]: yes, it is injected to the DeliveryService

File: cg/services/file_delivery/deliver_files_service/deliver_files_service_factory.py [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+            tags_fetcher=file_tag_fetcher,
+        )
+
+    @staticmethod
+    def _get_sample_file_formatter(
+        workflow: Workflow,
+    ) -> SampleFileFormatter | SampleFileConcatenationFormatter:
+        """Get the file formatter service based on the workflow."""
+        if workflow in [Workflow.MICROSALT, Workflow.MUTANT]:
+            return SampleFileConcatenationFormatter(FastqConcatenationService())
```
[the remark this answers is not in the document]
Comment (ChrOertlin) [reply cg/services/file_delivery/deliver_files_service/deliver_files_service_factory.py:88]: This way we can change the concatenation service if a different one is required easily.

**Pull request merged**